consensus/fork.go

```diff
@@ -16,95 +16,108 @@ func (s *State) heavierFork(newNode *BlockNode) bool {
+	s.currentBlockID = bn.Parent.Block.ID()
+}
 
-	// Update the CurrentBlock and CurrentPath variables of the longest fork.
-	delete(s.currentPath, s.height())
-	s.currentBlockID = s.currentBlock().ParentBlockID
+// rewindToNode will rewind blocks until `bn` is the highest block.
+func (s *State) rewindToNode(bn *BlockNode) (rewoundNodes []*BlockNode) {
+	// Sanity check  - make sure that bn is in the currentPath.
+	if DEBUG {
+		if bn.Block.ID() != s.currentPath[bn.Height] {
+			panic("bad use of rewindToNode")
+		}
+	}
+
+	// Remove blocks from the ConsensusState until we get to the
+	// same parent that we are forking from.
+	for s.currentBlockID != bn.Block.ID() {
+		rewoundNodes = append(rewoundNodes, s.currentBlockNode())
+		s.invertRecentBlock()
+	}
 	return
 }
 
 // s.integrateBlock() will verify the block and then integrate it into the
 // consensus state.
-func (s *State) integrateBlock(b Block, bd *BlockDiff) (diffs []OutputDiff, err error) {
-	bd.CatalystBlock = b.ID()
+func (s *State) generateAndApplyDiff(bn *BlockNode) (err error) {
+	// Sanity check - generate should only be called if the diffs have not yet
+	// been generated.
+	if DEBUG {
+		if bn.DiffsGenerated {
+			panic("misuse of generateAndApplyDiff")
+		}
+	}
+
+	// Update the current block and current path.
+	s.currentBlockID = bn.Block.ID()
+	s.currentPath[bn.Height] = bn.Block.ID()
 
-	var appliedTransactions []Transaction
-	minerSubsidy := Currency(0)
-	for _, txn := range b.Transactions {
+	minerSubsidy := CalculateCoinbase(s.height())
+	for _, txn := range bn.Block.Transactions {
 		err = s.validTransaction(txn)
 		if err != nil {
 			break
 		}
 
 		// Apply the transaction to the ConsensusState, adding it to the list of applied transactions.
-		transactionDiff := s.applyTransaction(txn)
-		appliedTransactions = append(appliedTransactions, txn)
-		diffs = append(diffs, transactionDiff.OutputDiffs...)
-		bd.TransactionDiffs = append(bd.TransactionDiffs, transactionDiff)
+		outputDiffs, contractDiffs := s.applyTransaction(txn)
+		bn.OutputDiffs = append(bn.OutputDiffs, outputDiffs...)
+		bn.ContractDiffs = append(bn.ContractDiffs, contractDiffs...)
 
 		// Add the miner fees to the miner subsidy.
 		for _, fee := range txn.MinerFees {
 			minerSubsidy += fee
 		}
 	}
-
 	if err != nil {
-		// Rewind transactions added.
-		for i := len(appliedTransactions) - 1; i >= 0; i-- {
-			s.invertTransaction(appliedTransactions[i])
-		}
+		s.invertRecentBlock()
 		return
 	}
 
 	// Perform maintanence on all open contracts.
-	diffSet := s.applyContractMaintenance(&bd.BlockChanges)
-	diffs = append(diffs, diffSet...)
-
-	// Update the current block and current path variables of the longest fork.
-	height := s.blockMap[b.ID()].Height
-	s.currentBlockID = b.ID()
-	s.currentPath[height] = b.ID()
+	outputDiffs, contractDiffs := s.applyContractMaintenance()
+	bn.OutputDiffs = append(bn.OutputDiffs, outputDiffs...)
+	bn.ContractDiffs = append(bn.ContractDiffs, contractDiffs...)
```

Comment:
this pattern pops up in a couple places. Would be nice if you could just write:
```go
bn.Diffs = append(bn.Diffs, s.applyContractMaintenance())
```
this might complicate the code in other places though.